### PR TITLE
Implemented new kernel interface for compositional properties

### PR DESCRIPTION
**Pull request by francoishamon**
This PR implements a simple kernel interface for the compositional property updates:
- Component fractions
- Phase volume fractions
- Phase mobilities (FV and mimetic)
- Total mass densities (wells)

with a base class that contains the various `launch` functions, and derived classes implementing the `compute` functions. The goal is to make the extension to thermal easier using callbacks in the `compute` functions. The non-local terms (fluxes, etc) will be done in #1696, and the well kernels in another PR.

The downside is that I am no longer able to use explicit instantiation with this approach.

The PR is ready for ready for review, and passes the integrated tests on Quartz and Lassen (no rebaseline necessary).

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp [hunk 1]
```diff
@@ -519,65 +519,20 @@ void CompositionalMultiphaseFVM::updatePhaseMobility( ObjectManagerBase & dataGr
...
-                                                dPhaseMob_dComp );
+  bool const isIsothermal = true;
+  PhaseMobilityKernelFactory::
+    createAndLaunch< parallelDevicePolicy<> >
+    ( isIsothermal,
+    m_numComponents,
+    m_numPhases,
+    dataGroup,
+    fluid,
+    relperm );
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp:535]: Not too thrilled about this formatting, i'm guessing uncrustify doesn't align arguments unless the first one is one the same line as function name? (namespace on separate line is fine though)

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.cpp [hunk 2]
```diff
@@ -17,302 +17,11 @@
...
-INST_PhaseVolumeFractionKernel( 1, 3 );
-INST_PhaseVolumeFractionKernel( 2, 3 );
-INST_PhaseVolumeFractionKernel( 3, 3 );
-INST_PhaseVolumeFractionKernel( 4, 3 );
-INST_PhaseVolumeFractionKernel( 5, 3 );
-
-#undef INST_PhaseVolumeFractionKernel
-
-} // namespace CompositionalMultiphaseBaseKernels
+{} // namespace CompositionalMultiphaseBaseKernels
```
Comment (klevzoff) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.cpp:25]: My guess would be we're not going back to explicit instantiations, so let's remove the empty files(s)?

**Review by klevzoff — APPROVED** [review]
LGTM!

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp [hunk 3]
```diff
@@ -41,94 +40,436 @@ using namespace constitutive;
...
+  arrayView2d< real64, compflow::USD_COMP > m_compFrac;
+  arrayView3d< real64, compflow::USD_COMP_DC > m_dCompFrac_dCompDens;
+
+};
+
+namespace internal
+{
+
+template< typename T, typename LAMBDA >
+void kernelLaunchSelectorCompSwitch( T value, LAMBDA && lambda )
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp]: I moved the switchyard above `ComponentFractionKernel`

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp [hunk 4]
```diff
@@ -41,94 +40,436 @@ using namespace constitutive;
...
+  constexpr void
+  operator()( Ts && ... ) const {}
+};
+
+/**
+ * @class PropertyKernelBase
+ * @tparam NUM_COMP number of fluid components
+ * @brief Define the base interface for the property update kernels
+ */
+template< localIndex NUM_COMP >
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp]: I updated all the component and phase templates (`localIndex` -> `integer`) in all the compositional flow kernels. I will do the well kernels in a separate PR

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp [hunk 5]
```diff
@@ -41,94 +40,436 @@ using namespace constitutive;
...
   static void
-  launch( localIndex const size,
-          arrayView2d< real64 const, compflow::USD_COMP > const & compDens,
-          arrayView2d< real64 const, compflow::USD_COMP > const & dCompDens,
-          arrayView2d< real64, compflow::USD_COMP > const & compFrac,
-          arrayView3d< real64, compflow::USD_COMP_DC > const & dCompFrac_dCompDens );
+  launch( localIndex const numElems,
+          KERNEL_TYPE const & kernelComponent )
+  {
+    GEOSX_MARK_FUNCTION;
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp]: You are right, I removed these `GEOSX_MARK_FUNCTION` markers

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp [hunk 6]
```diff
@@ -41,94 +40,436 @@ using namespace constitutive;
...
-          arrayView2d< real64, compflow::USD_PHASE > const & dPhaseVolFrac_dPres,
-          arrayView3d< real64, compflow::USD_PHASE_DC > const & dPhaseVolFrac_dComp );
+  createAndLaunch( localIndex const numComp,
+                   ObjectManagerBase & subRegion )
+  {
+    internal::kernelLaunchSelectorCompSwitch( numComp, [&] ( auto NC )
+    {
+      localIndex constexpr NUM_COMP = NC();
+      ComponentFractionKernel< NUM_COMP > kernel( subRegion );
+      ComponentFractionKernel< NUM_COMP >::template launch< POLICY, ComponentFractionKernel< NUM_COMP > >( subRegion.size(), kernel );
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp]: Yes, deducing the second template argument seems to work, that is great, thanks!

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp [hunk 7]
```diff
@@ -41,94 +40,436 @@ using namespace constitutive;
...
+   * @brief Compute the property in an element
+   * @tparam FUNC the type of the function that can be used to customize the kernel
+   * @param[in] ei the element index
+   * @param[in] phaseVolFractionKernelOp the function used to customize the kernel
+   */
+  template< typename FUNC = NoOpFunc >
+  GEOSX_HOST_DEVICE
+  void compute( localIndex const ei,
+                FUNC && kernelOp = NoOpFunc{} ) const
+  { GEOSX_UNUSED_VAR( ei, kernelOp ); }
```
[the remark this answers is not in the document]
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.hpp]: I removed these functions (constructor and `compute` function of the base class)

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp (hunk 1 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp]: This is fixed, now if looks like that:
https://github.com/GEOSX/GEOSX/blob/2ae37eed2e17c251e35bca817c427d563dd7b85f/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp#L947-L956

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.cpp (hunk 2 above)
Comment (francoishamon) [reply src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBaseKernels.cpp]: I removed two empty files in the last commit: `SinglePhaseBaseKernels.cpp` and `CompositionalMultiphaseBaseKernels.cpp`

**Comment by mkhait** [comment]
"The downside is that I am no longer able to use explicit instantiation with this approach"

@francoishamon, why is that downside? I always thought its a benefit )

**Comment by francoishamon** [comment]
Previously, in the compositional solver, we were explicitly instantiating the templated functions (mainly the `launch` functions) in the cpp files to speed up compilation (particularly for the GPU builds that take forever). But now, with the new kernel classes, it has become harder to do and I removed many of these explicit template instantiations. 

I don't know enough about compilation to provide a more detailed answer :) and experts will jump in to add details. The long term approach that we want to follow is to use just-in-time compilation to speed up our builds, as in this [PR](https://github.com/GEOSX/GEOSX/pull/1333).

**Pull request merged**